BF - Fix io_orientation to process input axes by strength

leoyala opened the pull request:
This PR introduces changes aiming to make the axis assignment based on image affines more consistent. Instead of iterating over the axis in order `range(p)`, the axis are now iterated over based on the `strongest` axis according to the rotation-shear matrix (RS). 

Alternative implementation:
Add a parameter to `io_orientation` called `sort_by_strength` which would activate this new sorting method, while preserving the old behavior. This parameter should then be propagated to `as_closest_canonical` and other places that might need it. In my opinion this might introduce more complexity and make it harder to maintain. 

Notes:
In principle the sorting could be done based on the rotation matrix `R` instead of `RS`. However this causes `test_io_orientation` to fail due to the check when additional columns are added to the affine.

Closes #1449

leoyala committed:
BF - Fix io_orientation to process input axes by strength, ensuring consistent labeling. Add regression test for handling competing axes in orientation determination.

leoyala commented:
After testing the new implementation on a few hundred images that have non-standard orientations (e.g. oblique planes). The previous bug seems to be resolved for most cases, however I still found one image for which loading the image assigns what should be an Axial plane to the Sagittal one. 
The affine of this image is:
```
[
[ 1.94222772e+00, -1.87633261e-01, -2.54019409e-01, 1.09614174e+02],
[ 1.56332731e+00,  3.34649354e-01, -7.97063559e-02, -1.08926643e+02],
[ 2.16188765e+00, -7.34265447e-02,  2.85847723e-01, -8.57995453e+01],
[ 0.00000000e+00,  0.00000000e+00,  0.00000000e+00, 1.00000000e+00]
]
```
Even more, the old implementation (using `range(p)`) seems to load this image correctly.
Interestingly, running sflhd on this particular image returns the following, which seems to be the same as the new implementation.
```bash
sto_xyz:1	1.942228 -0.187633 -0.254019 109.614174 
sto_xyz:2	1.563327 0.334649 -0.079706 -108.926643 
sto_xyz:3	2.161888 -0.073427 0.285848 -85.799545 
sto_xyz:4	0.000000 0.000000 0.000000 1.000000 
sform_xorient	Left-to-Right
sform_yorient	Posterior-to-Anterior
sform_zorient	Inferior-to-Superior
``` 
Loading this image with another tool, like MITK, does seem to load the image in the correct orientation, assigning the axial plane to what the new implementation here assigns the sagittal one. 
Any suggestions for improvement are welcomed :)

leoyala committed:
BF - Refactors io_orientation to use rotation matrix for axes sorting. Updates regression test to verify orientation handling.

leoyala commented on nibabel/orientations.py:
```diff
@@ -75,7 +75,10 @@ def io_orientation(affine, tol=None):
-    for in_ax in range(p):
+    # Process input axes from strongest to weakest (stable on ties) so a given
+    # dimension is labeled consistently regardless of the original order.
+    in_axes = np.argsort(np.min(-(RS**2), axis=0), kind='stable')
```
I have pushed changes to address the rest of your comments. Should we wait for the comments from matthew before addressing the changes to the failing test? 
P.S.: I will be away for several weeks so it might take me a while to address those comments.

effigies committed:
test: Make assertions consistent, add more detail to comments

effigies committed:
Merge branch 'master' into fix-axis-orientation-canonical

effigies reviewed (approved):
Thanks for following up on this. I lost it in my holiday notifications.

effigies committed:
test: Fix variable name

effigies committed:
test: Update expectation for dropped columns

This is an unintended change in the test, but it arises in a far corner
case where columns differ by 10*floating point epsilon.
The labeling on degenerate affines is a change I'm willing to live with.

effigies commented:
@leoyala Please feel free to add yourself to the `.zenodo.json` file and `doc/source/index.rst`.
